### PR TITLE
[1.1] [project] train imagenet using large batch size

### PR DESCRIPTION
**Pull request by typhoonzero**
1. [Feature] multi_batch_merge pass, enable simulated multi gpu training on a single gpu to enlarge the total batch size
2. [Performance] fuse broadcast op by @Yancey1989 
3. [Feature] lars_momentum_op (new layer API)

### CODE REVIEW

**Comment by typhoonzero** [comment]
@panyx0718 Need api approval for this PR, thanks!

File: paddle/fluid/framework/ir/graph.h [hunk 1]
```diff
@@ -160,6 +161,15 @@ class Graph {
     return nullptr;
   }
 
+  void ResetNodeId() { Node::ResetId(); }
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/graph.h]: Please don't use ResetId(), it's a hack that was only used in inference

Comment (panyx0718) [inline paddle/fluid/framework/ir/graph.h]: Please don't use ResetId(), it's a hack that was only used in inference

Comment (panyx0718) [inline paddle/fluid/framework/ir/graph.h]: Please don't use ResetId(), it's a hack that was only used in inference

Comment (panyx0718) [inline paddle/fluid/framework/ir/graph.h]: !!

Comment (panyx0718) [inline paddle/fluid/framework/ir/graph.h]: !!

Comment (typhoonzero) [reply paddle/fluid/framework/ir/graph.h]: Done.

File: paddle/fluid/API.spec [hunk 2]
```diff
@@ -16,6 +16,7 @@ paddle.fluid.Executor.run ArgSpec(args=['self', 'program', 'feed', 'fetch_list',
 paddle.fluid.global_scope ArgSpec(args=[], varargs=None, keywords=None, defaults=None)
 paddle.fluid.scope_guard ArgSpec(args=[], varargs='args', keywords='kwds', defaults=None)
 paddle.fluid.DistributeTranspiler.__init__ ArgSpec(args=['self', 'config'], varargs=None, keywords=None, defaults=(None,))
+paddle.fluid.DistributeTranspiler.append_bn_repeat_init_op ArgSpec(args=['main_prog', 'startup_prog', 'num_repeats'], varargs=None, keywords=None, defaults=None)
```
Comment (panyx0718) [inline paddle/fluid/API.spec]: not a good api...

File: paddle/fluid/API.spec [hunk 3]
```diff
@@ -312,6 +313,7 @@ paddle.fluid.contrib.QuantizeTranspiler.convert_to_int8 ArgSpec(args=['self', 'p
 paddle.fluid.contrib.QuantizeTranspiler.freeze_program ArgSpec(args=['self', 'program', 'place', 'fuse_bn', 'scope'], varargs=None, keywords=None, defaults=(False, None))
 paddle.fluid.contrib.QuantizeTranspiler.training_transpile ArgSpec(args=['self', 'program', 'startup_program'], varargs=None, keywords=None, defaults=(None, None))
 paddle.fluid.transpiler.DistributeTranspiler.__init__ ArgSpec(args=['self', 'config'], varargs=None, keywords=None, defaults=(None,))
+paddle.fluid.transpiler.DistributeTranspiler.append_bn_repeat_init_op ArgSpec(args=['main_prog', 'startup_prog', 'num_repeats'], varargs=None, keywords=None, defaults=None)
```
Comment (panyx0718) [inline paddle/fluid/API.spec]: not a good api...

File: paddle/fluid/API.spec [hunk 4]
```diff
@@ -353,6 +355,8 @@ paddle.fluid.optimizer.ModelAverage.__init__ ArgSpec(args=['self', 'average_wind
 paddle.fluid.optimizer.ModelAverage.apply ArgSpec(args=[], varargs='args', keywords='kwds', defaults=None)
 paddle.fluid.optimizer.ModelAverage.minimize ArgSpec(args=['self', 'loss', 'startup_program', 'parameter_list', 'no_grad_set'], varargs=None, keywords=None, defaults=(None, None, None))
 paddle.fluid.optimizer.ModelAverage.restore ArgSpec(args=['self', 'executor'], varargs=None, keywords=None, defaults=None)
+paddle.fluid.optimizer.LarsMomentumOptimizer.__init__ ArgSpec(args=['self', 'learning_rate', 'momentum', 'lars_coeff', 'lars_weight_decay', 'regularization', 'name'], varargs=None, keywords=None, defaults=(0.001, 0.0005, None, None))
```
Comment (panyx0718) [inline paddle/fluid/API.spec:357]: put this 2 in contrib?

File: paddle/fluid/operators/lars_momentum_op.cc [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+#include "paddle/fluid/operators/momentum_op.h"
+
+namespace paddle {
+namespace operators {
+
+class LarsMomentumOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() override {
+    AddInput("Param",
+             "(Tensor, default Tensor<float>) "
```
Comment (chengduoZH) [inline paddle/fluid/operators/lars_momentum_op.cc]: `Tensor ` -> `LoDTensor`

File: paddle/fluid/operators/lars_momentum_op.cc [hunk 6]
```diff
@@ -0,0 +1,79 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserved.
```
Comment (chengduoZH) [inline paddle/fluid/operators/lars_momentum_op.cc]: `2016`->`2018`

File: paddle/fluid/framework/ir/graph.cc [hunk 7]
```diff
@@ -89,31 +89,34 @@ bool IsDistTrainOp(ir::Node *node, const std::vector<std::string> &send_vars,
 Graph::Graph(const ProgramDesc &program) : program_(program) {
   // Make the nodes id start from 0.
   Node::ResetId();
+  InitFromProgram(program_);
+  ResolveHazard(var_nodes_);
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/graph.cc]: Please don't make `var_nodes_` a class member

File: paddle/fluid/framework/parallel_executor.cc [hunk 8]
```diff
@@ -156,6 +147,15 @@ ParallelExecutor::ParallelExecutor(
                            params, member_->local_scopes_, member_->use_cuda_);
 #endif
 
+  // Step 3. Create vars in each scope. Passes may also create new vars.
+  std::vector<details::VariableInfo> var_infos;
+  for (auto *var : graph->AllVars()) {
```
Comment (panyx0718) [inline paddle/fluid/framework/parallel_executor.cc]: Please don't use AllVars. Search through the graph and find the "current" vars.

File: paddle/fluid/operators/lars_momentum_op.cc [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+)DOC");
+  }
+};
+}  // namespace operators
+}  // namespace paddle
+
+namespace ops = paddle::operators;
+REGISTER_OP_WITHOUT_GRADIENT(lars_momentum, ops::MomentumOp,
+                             ops::LarsMomentumOpMaker);
```
Comment (chengduoZH) [inline paddle/fluid/operators/lars_momentum_op.cc]: `REGISTER_OP_WITHOUT_GRADIENT` can be replaced by `REGISTER_OPERATOR`.

File: paddle/fluid/framework/ir/graph.h [hunk 10]
```diff
@@ -175,6 +183,8 @@ class Graph {
   std::map<std::string, std::function<void(void)>> attr_dels_;
   std::map<ir::Node *, std::unique_ptr<ir::Node>> nodes_;
   std::unordered_set<ir::Node *> node_set_;
+  // var nodes for each var name, will have multiple versions in SSA
+  std::map<std::string, std::vector<ir::Node *>> var_nodes_;
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/graph.h]: Don't make this a member. You are not making it up-to-date.

File: paddle/fluid/framework/ir/graph.h [hunk 11]
```diff
@@ -160,6 +161,13 @@ class Graph {
     return nullptr;
   }
 
+  void InitFromProgram(const ProgramDesc &program);
+
+  void ResolveHazard(
+      const std::map<std::string, std::vector<ir::Node *>> &var_nodes);
+
+  std::vector<VarDesc *> AllVars();
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/graph.h]: Don't implement it this way. If you want AllVars, add a graph_helper function.

File: paddle/fluid/framework/ir/multi_batch_merge_pass.cc [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+
+  result.ResolveHazard(created);
+  return graph;
+}
+
+}  // namespace ir
+}  // namespace framework
+}  // namespace paddle
+
+REGISTER_PASS(multi_batch_merge_pass, paddle::framework::ir::BatchMergePass)
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/multi_batch_merge_pass.cc:314]: This pass needs documentation. It's not easy to know its meaning from the name.

File: paddle/fluid/operators/lars_momentum_op.cc [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+This optimizer use LARS (https://arxiv.org/abs/1708.03888) to optimize each
+weight using a local learning rate:
+
+$$
+learning\_rate *= lars_coeff * sqrt(sumsq(param)) 
+    / (sqrt(sumsq(gradient))+ lars\_weight\_decay * sqrt(sumsq(param))) \\
+velocity = mu * velocity + 
+    (gradient + lars\_weight\_decay * param) \\
+param = param - learning\_rate * velocity. \\
+$$
```
Comment (chengduoZH) [inline paddle/fluid/operators/lars_momentum_op.cc:64]: The format of this expression should be refine.
lar

File: paddle/fluid/framework/ir/multi_batch_merge_pass.cc [hunk 14]
```diff
@@ -0,0 +1,317 @@
...
+    if (node->IsVar()) continue;
+    int op_role = boost::get<int>(node->Op()->GetAttr(
+        framework::OpProtoAndCheckerMaker::OpRoleAttrName()));
+    if (op_role == static_cast<int>(framework::OpRole::kForward) ||
+        op_role == static_cast<int>(framework::OpRole::kBackward) ||
+        op_role == static_cast<int>(framework::OpRole::kLoss) ||
+        op_role == (static_cast<int>(framework::OpRole::kForward) |
+                    static_cast<int>(framework::OpRole::kLoss)) ||
+        op_role == (static_cast<int>(framework::OpRole::kBackward) |
+                    static_cast<int>(framework::OpRole::kLoss))) {
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/multi_batch_merge_pass.cc]: Use & to make this simpler?

File: paddle/fluid/framework/ir/multi_batch_merge_pass.cc [hunk 15]
```diff
@@ -0,0 +1,317 @@
...
+      forward_backward_ops.push_back(node);
+    } else if (op_role == static_cast<int>(framework::OpRole::kOptimize) ||
+               op_role == static_cast<int>(framework::OpRole::kDist) ||
+               op_role == static_cast<int>(framework::OpRole::kRPC)) {
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/multi_batch_merge_pass.cc]: What about `kOptimize | kLrSched`

File: paddle/fluid/operators/lars_momentum_op.cc [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include "paddle/fluid/operators/lars_momentum_op.h"
+#include "paddle/fluid/operators/momentum_op.h"
+
+namespace paddle {
+namespace operators {
+
+class LarsMomentumOpMaker : public framework::OpProtoAndCheckerMaker {
```
Comment (chengduoZH) [inline paddle/fluid/operators/lars_momentum_op.cc:21]: Does `LarsMomentumOp` support sparse gradient?

File: paddle/fluid/framework/ir/multi_batch_merge_pass.cc [hunk 17]
```diff
@@ -0,0 +1,317 @@
...
+      optimize_ops.push_back(node);
+      auto op_role_var = node->Op()->GetNullableAttr(
+          OpProtoAndCheckerMaker::OpRoleVarAttrName());
+      auto op_role_vars = boost::get<std::vector<std::string>>(op_role_var);
+      for (size_t i = 0; i < op_role_vars.size(); i += 2) {
+        grad_names.insert(op_role_vars[i + 1]);
+      }
+    } else if (op_role == static_cast<int>(framework::OpRole::kLRSched)) {
+      lr_ops.push_back(node);
+    }
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/multi_batch_merge_pass.cc:107]: else LOG(FATAL)?

File: paddle/fluid/operators/lars_momentum_op.cu [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+    v_out[i] = v_new;
+    p_out[i] = p[i] - v_new;
+  }
+}
+
+template <typename DeviceContext, typename T>
+class LarsMomentumOpCUDAKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    auto param_out = ctx.Output<framework::Tensor>("ParamOut");
```
Comment (chengduoZH) [inline paddle/fluid/operators/lars_momentum_op.cu]: We had better use `ctx.Output<framework::LoDTensor>` here.

File: paddle/fluid/framework/ir/multi_batch_merge_pass.cc [hunk 19]
```diff
@@ -0,0 +1,317 @@
...
+    std::unordered_set<ir::Node*> copied;
+    for (size_t node_idx = 0; node_idx < forward_backward_ops.size();
+         ++node_idx) {
+      auto node = forward_backward_ops[node_idx];
+      OpDesc repeated_op(*(node->Op()), node->Op()->Block());
+      // 3. rename grad outputs to current repeat.
+      for (auto outname : repeated_op.OutputArgumentNames()) {
+        if (grad_names.find(outname) != grad_names.end()) {
+          std::string new_gname = string::Sprintf("%s.repeat.%d", outname, i);
+          repeated_op.RenameOutput(outname, new_gname);
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/multi_batch_merge_pass.cc:126]: why do you need rename? I think you only need to make the graph connection right.

File: paddle/fluid/framework/ir/multi_batch_merge_pass.cc [hunk 20]
```diff
@@ -0,0 +1,317 @@
...
+                                 new_mean_name);
+        repeated_op.RenameOutput(repeated_op.Output("VarianceOut")[0],
+                                 new_var_name);
+      }
+
+      // 3.9 do copy
+      auto repeated_node = result.CreateOpNode(&repeated_op);
+      copied.insert(node);
+
+      // 4. add deps between repeats
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/multi_batch_merge_pass.cc:154]: feels dangerous. any tests?

File: paddle/fluid/framework/ir/multi_batch_merge_pass.cc [hunk 21]
```diff
@@ -0,0 +1,317 @@
...
+      if (node_idx == 0 && prev_repeat_last_op_node) {
+        auto* depvar = result.CreateControlDepVar();
+        prev_repeat_last_op_node->outputs.push_back(depvar);
+        depvar->inputs.push_back(prev_repeat_last_op_node);
+        repeated_node->inputs.push_back(depvar);
+        depvar->outputs.push_back(repeated_node);
+      }
+
+      for (auto in_node : node->inputs) {
+        if (in_node->IsCtrlVar()) {
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/multi_batch_merge_pass.cc:167]: why skip control var? control var is not copied? how do you deal with control var deps between forward/backward and optimize ops?

File: paddle/fluid/operators/lars_momentum_op.cu [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+                 (g_norm[0] + lars_weight_decay * p_norm[0]);
+    }
+    T v_new = v[i] * mu + local_lr * (g[i] + lars_weight_decay * p[i]);
+    v_out[i] = v_new;
+    p_out[i] = p[i] - v_new;
+  }
+}
+
+template <typename DeviceContext, typename T>
+class LarsMomentumOpCUDAKernel : public framework::OpKernel<T> {
```
Comment (panyx0718) [inline paddle/fluid/operators/lars_momentum_op.cu:42]: Can share some code with MomentumOp?

File: paddle/fluid/pybind/pybind.cc [hunk 23]
```diff
@@ -817,6 +817,10 @@ All parameter, weight, gradient are variables in Paddle.
           [](BuildStrategy &self, bool b) {
             self.enable_data_balance_ = b;
           })  // FIXME(chengudo): enable_data_balance seems not important
+      .def_property(
+          "multi_batch_merge_repeats",
```
Comment (panyx0718) [inline paddle/fluid/pybind/pybind.cc]: I suggest you example in test_pass_builder.py to insert a pass and avoid this API. Very few people use it.

File: python/paddle/fluid/tests/unittests/dist_mnist.py [hunk 24]
```diff
@@ -90,12 +90,11 @@ def get_model(self, batch_size=2):
 
         inference_program = fluid.default_main_program().clone()
         # Optimization
-        opt = fluid.optimizer.AdamOptimizer(
-            learning_rate=0.001, beta1=0.9, beta2=0.999)
+        opt = fluid.optimizer.Momentum(learning_rate=0.001, momentum=0.9)
```
Comment (panyx0718) [inline python/paddle/fluid/tests/unittests/dist_mnist.py]: why change here?

File: python/paddle/fluid/transpiler/distribute_transpiler.py [hunk 25]
```diff
@@ -1430,7 +1430,7 @@ def _get_optimizer_input_shape(self, op_type, varkey, orig_shape,
         elif op_type == "adamax":
             if varkey in ["Moment", "InfNorm"]:
                 return param_shape
-        elif op_type == "momentum":
+        elif op_type in ["momentum", "lars_momentum"]:
```
Comment (panyx0718) [inline python/paddle/fluid/transpiler/distribute_transpiler.py:1434]: add a else below to raise an exception?

File: python/paddle/fluid/transpiler/distribute_transpiler.py [hunk 26]
```diff
@@ -1795,3 +1795,65 @@ def _get_optimize_pass(self):
             else:
                 pass
         return opt_ops, params_grads
+
+    @staticmethod
+    def append_bn_repeat_init_op(main_prog, startup_prog, num_repeats):
```
Comment (panyx0718) [inline python/paddle/fluid/transpiler/distribute_transpiler.py]: Is there any way to avoid this API? This is a static method. Can you put it in contrib?

File: paddle/fluid/API.spec (hunk 2 above)
Comment (typhoonzero) [reply paddle/fluid/API.spec]: I know, for now we cannot change startup program when adding a pass

File: paddle/fluid/API.spec (hunk 4 above)
Comment (typhoonzero) [reply paddle/fluid/API.spec:357]: Seems we don't have contrib struct for operators, this may take some time. And, this operator is now quiet commonly used in CV.

File: python/paddle/fluid/transpiler/distribute_transpiler.py (hunk 26 above)
Comment (typhoonzero) [reply python/paddle/fluid/transpiler/distribute_transpiler.py]: Not currently until we find a way to modify startup program in passes.

File: paddle/fluid/framework/ir/multi_batch_merge_pass.cc (hunk 19 above)
Comment (typhoonzero) [reply paddle/fluid/framework/ir/multi_batch_merge_pass.cc:126]: parallel executor need OpDesc to run.

File: paddle/fluid/framework/ir/multi_batch_merge_pass.cc (hunk 21 above)
Comment (typhoonzero) [reply paddle/fluid/framework/ir/multi_batch_merge_pass.cc:167]: Control var is not copied, all WAR, WAR control var will be rebuilt by calling `graph->ResolveHazard`

File: paddle/fluid/operators/lars_momentum_op.cu (hunk 22 above)
Comment (typhoonzero) [reply paddle/fluid/operators/lars_momentum_op.cu:42]: The code is simple and not quite the same. I'll try cleanup this later.

File: paddle/fluid/API.spec (hunk 2 above)
Comment (panyx0718) [reply paddle/fluid/API.spec]: but this api is so ugly that I want to this to be a public api

Comment (typhoonzero) [reply paddle/fluid/API.spec]: Should have a way to let user make full use of multi batch merge, or user have to write it in their own code: https://github.com/typhoonzero/paddle_imagenet_fast/blob/master/fast_imagenet_train.py#L165

Comment (typhoonzero) [reply paddle/fluid/API.spec]: I'll put this function in some demo code util we find some ways to change startup program in a pass

File: paddle/fluid/pybind/pybind.cc (hunk 23 above)
Comment (typhoonzero) [reply paddle/fluid/pybind/pybind.cc]: Tried, but failed, this pass must run before multi_device_graph_pass

**Comment by typhoonzero** [comment]
@panyx0718 @chengduoZH comments done, those places I didn't update I've put some reason there.

File: paddle/fluid/framework/ir/graph.h [hunk 27]
```diff
@@ -63,6 +63,7 @@ namespace ir {
 class Graph {
  public:
   explicit Graph(const ProgramDesc &program);
+  Graph() {}
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/graph.h]: why is this added?

File: paddle/fluid/framework/parallel_executor.cc [hunk 28]
```diff
@@ -156,6 +147,19 @@ ParallelExecutor::ParallelExecutor(
                            params, member_->local_scopes_, member_->use_cuda_);
 #endif
 
+  // Step 3. Create vars in each scope. Passes may also create new vars.
+  //         skip control vars and empty vars
+  std::vector<details::VariableInfo> var_infos;
+  for (auto &node : graph->Nodes()) {
+    if (node->IsVar() && !node->IsCtrlVar() && !node->Name().empty() &&
+        node->Var()) {
```
Comment (panyx0718) [inline paddle/fluid/framework/parallel_executor.cc]: I think we don't allow name to be empty. We can enforce this somewhere.

File: python/paddle/fluid/optimizer.py [hunk 29]
```diff
@@ -398,6 +399,91 @@ def _append_optimize_op(self, block, param_and_grad):
...
+        lars_weight_decay (float): weight decay coefficient for decaying using LARS.
+        regularization: A Regularizer, such as
+                        fluid.regularizer.L2DecayRegularizer.
+        name: A optional name prefix.
+        
+
+    Examples:
+        .. code-block:: python
+
+            optimizer = fluid.optimizer.Momentum(learning_rate=0.2, momentum=0.1)
```
Comment (panyx0718) [inline python/paddle/fluid/optimizer.py]: update?

File: python/paddle/fluid/tests/unittests/dist_mnist.py [hunk 30]
```diff
@@ -95,7 +95,7 @@ def get_model(self, batch_size=2):
 
         # Reader
         train_reader = paddle.batch(
-            paddle.dataset.mnist.train(), batch_size=batch_size)
+            paddle.dataset.mnist.test(), batch_size=batch_size)
```
Comment (panyx0718) [inline python/paddle/fluid/tests/unittests/dist_mnist.py:98]: why change here? Should it be renamed test_reader?

File: python/paddle/fluid/tests/unittests/test_dist_mnist_batch_merge.py [hunk 31]
```diff
@@ -0,0 +1,67 @@
...
+
+        batch_merge_losses = self._run_local(
+            model_file,
+            required_envs,
+            check_error_log=check_error_log,
+            batch_size=2,
+            batch_merge_repeat=2)
+        # Ensure both result have values.
+        self.assertGreater(len(no_merge_losses), 1)
+        self.assertEqual(len(no_merge_losses), len(batch_merge_losses))
```
Comment (panyx0718) [inline python/paddle/fluid/tests/unittests/test_dist_mnist_batch_merge.py:63]: should the loss be equal?

File: paddle/fluid/pybind/pybind.cc (hunk 23 above)
Comment (panyx0718) [reply paddle/fluid/pybind/pybind.cc]: It should be simple. `all_passes` is a list of pass names. You just need to find the position of `multi_device_graph_pass`, and insert your pass at that position. You can debug by printing all passes. You can update test_pass_builder.py to try it.
```
pass_builder.insert_pass(
            len(pass_builder.all_passes()), "graph_viz_pass")
        self.assertEqual(origin_len + 2, len(pass_builder.all_passes()))
```

File: paddle/fluid/framework/op_proto_maker.h [hunk 32]
```diff
@@ -25,12 +25,12 @@ enum class OpRole {
...
   // RPC role is for send/recv releated op
-  kRPC = 0x0003,
+  kRPC = 0x0004,
   // Dist role is for split_byref/split_selected_rows/concat
   // used for distributed training.
-  kDist = 0x0004,
+  kDist = 0x0008,
   // Tag all learning rate scheduler operators.
-  kLRSched = 0x0005,
+  kLRSched = 0x0016,
```
Comment (panyx0718) [inline paddle/fluid/framework/op_proto_maker.h]: this fix should probably be checked in separately ASAP.

File: paddle/fluid/framework/ir/multi_batch_merge_pass.cc [hunk 33]
```diff
@@ -0,0 +1,315 @@
...
+        framework::OpProtoAndCheckerMaker::OpRoleAttrName()));
+    if ((op_role == static_cast<int>(framework::OpRole::kForward)) ||
+        (op_role & static_cast<int>(framework::OpRole::kBackward)) ||
+        (op_role & static_cast<int>(framework::OpRole::kLoss))) {
+      forward_backward_ops.push_back(node);
+    } else if ((op_role & static_cast<int>(framework::OpRole::kOptimize)) ||
+               (op_role & static_cast<int>(framework::OpRole::kDist)) ||
+               (op_role & static_cast<int>(framework::OpRole::kRPC))) {
+      optimize_ops.push_back(node);
+      auto op_role_var = node->Op()->GetNullableAttr(
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/multi_batch_merge_pass.cc:97]: should check empty?

File: paddle/fluid/framework/ir/multi_batch_merge_pass.cc [hunk 34]
```diff
@@ -0,0 +1,315 @@
...
+
+      // 3.9 do copy
+      auto repeated_node = result.CreateOpNode(&repeated_op);
+      copied.insert(node);
+
+      // 4. add deps between repeats
+      if (node_idx == forward_backward_ops.size() - 1) {
+        prev_repeat_last_op_node = repeated_node;
+      }
+      if (node_idx == 0 && prev_repeat_last_op_node) {
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/multi_batch_merge_pass.cc:158]: prev_repeat_last_op_node is not always the last executed op. so the dependency might not always work. You should find all ops with no follow up ops

File: paddle/fluid/framework/ir/multi_batch_merge_pass.cc [hunk 35]
```diff
@@ -0,0 +1,315 @@
...
+      // 3. rename grad outputs to current repeat.
+      for (auto outname : repeated_op.OutputArgumentNames()) {
+        if (grad_names.find(outname) != grad_names.end()) {
+          std::string new_gname = string::Sprintf("%s.repeat.%d", outname, i);
+          repeated_op.RenameOutput(outname, new_gname);
+        }
+      }
+      // 3.5 let batch_norm ops use independent vars, note batch_norm_grad do
+      // not need this update
+      if (node->Name() == "batch_norm") {
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/multi_batch_merge_pass.cc:131]: how are you dealing with fetch_op?

File: paddle/fluid/framework/ir/multi_batch_merge_pass.cc [hunk 36]
```diff
@@ -0,0 +1,315 @@
...
+
+      for (auto in_node : node->inputs) {
+        if (in_node->IsCtrlVar()) {
+          continue;
+        }
+        ir::Node* var = nullptr;
+        auto updated_var = UpdateGradVarDesc(in_node->Var(), i, grad_names,
+                                             bn_vars_need_rename);
+        // should be initialized by startup, how to initilize tensor in the
+        // scope?
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/multi_batch_merge_pass.cc:174]: Can you do some TensorCopy?

File: paddle/fluid/framework/ir/multi_batch_merge_pass.cc (hunk 35 above)
Comment (panyx0718) [inline paddle/fluid/framework/ir/multi_batch_merge_pass.cc:131]: this is a bit hard-code. In the future, there might be batch_norm2, like reshape2

File: paddle/fluid/framework/ir/multi_batch_merge_pass.cc [hunk 37]
```diff
@@ -0,0 +1,315 @@
...
+    for (auto node : lr_ops) {
+      copy_node(node);
+    }
+    for (auto node : optimize_ops) {
+      copy_node(node);
+    }
+  }
+
+  result.ResolveHazard(created);
+  return graph;
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/multi_batch_merge_pass.cc:307]: You might want to check GraphNum?

**Review by panyx0718 — COMMENTED** [review]
has the merge_pass been tested in a model?

**Review by panyx0718 — COMMENTED** [review]
where is fused_broadcast tested?

**Comment by panyx0718** [comment]
what's the experiment result of fuse_broadcast? multi_merge_pass? or lars_momentum?

File: python/paddle/fluid/tests/unittests/dist_mnist.py (hunk 30 above)
Comment (typhoonzero) [reply python/paddle/fluid/tests/unittests/dist_mnist.py:98]: To keep test result fixed.

File: paddle/fluid/framework/ir/multi_batch_merge_pass.cc (hunk 35 above)
Comment (typhoonzero) [reply paddle/fluid/framework/ir/multi_batch_merge_pass.cc:131]: I'll keep this here, no need add any tags or roles for batch_norm op.

File: paddle/fluid/framework/ir/multi_batch_merge_pass.cc (hunk 36 above)
Comment (typhoonzero) [reply paddle/fluid/framework/ir/multi_batch_merge_pass.cc:174]: Can do. will add in next refine.

File: paddle/fluid/framework/ir/multi_batch_merge_pass.cc (hunk 35 above)
Comment (typhoonzero) [reply paddle/fluid/framework/ir/multi_batch_merge_pass.cc:131]: What's the case? you mean people may fetch the original output of batch_norm?

File: paddle/fluid/framework/ir/multi_batch_merge_pass.cc (hunk 34 above)
Comment (typhoonzero) [reply paddle/fluid/framework/ir/multi_batch_merge_pass.cc:158]: For most cases, last backward op always depend on the previous grad output (chain rule), and if  there are cases not like this, and that case can use this feature for speed up, we can refine this.

File: paddle/fluid/framework/ir/multi_batch_merge_pass.cc (hunk 33 above)
Comment (typhoonzero) [reply paddle/fluid/framework/ir/multi_batch_merge_pass.cc:97]: Empty list will not going into the for loop

File: paddle/fluid/framework/op_proto_maker.h (hunk 32 above)
Comment (typhoonzero) [reply paddle/fluid/framework/op_proto_maker.h]: The fix is also in https://github.com/PaddlePaddle/Paddle/pull/14072

File: python/paddle/fluid/tests/unittests/test_dist_mnist_batch_merge.py (hunk 31 above)
Comment (typhoonzero) [reply python/paddle/fluid/tests/unittests/test_dist_mnist_batch_merge.py:63]: This is a todo, some distributed test cases still didn't match for now.

File: paddle/fluid/framework/ir/multi_batch_merge_pass.cc (hunk 37 above)
Comment (typhoonzero) [reply paddle/fluid/framework/ir/multi_batch_merge_pass.cc:307]: GraphNum is checked at `parallel_executor.cc`, see: https://github.com/PaddlePaddle/Paddle/pull/14072

File: paddle/fluid/framework/op_proto_maker.h (hunk 32 above)
Comment (typhoonzero) [reply paddle/fluid/framework/op_proto_maker.h]: `multi_batch_merge_repeats` done.

If we expose pass builder to users, must find a way to expose passes' document.

**Comment by typhoonzero** [comment]
@panyx0718 The model config is at https://github.com/PaddlePaddle/models/blob/develop/fluid/image_classification/models/resnet_dist.py. the result detail is at our inner wiki. The result is: using batch merge can achieve the same accuracy as when not using it, model config ref: https://arxiv.org/pdf/1706.02677.pdf

File: paddle/fluid/framework/ir/multi_batch_merge_pass.cc (hunk 35 above)
Comment (panyx0718) [reply paddle/fluid/framework/ir/multi_batch_merge_pass.cc:131]: I'm just guessing you might copy feed/fetch ops several times if they are forward op.

File: paddle/fluid/framework/ir/multi_batch_merge_pass.cc (hunk 37 above)
Comment (panyx0718) [reply paddle/fluid/framework/ir/multi_batch_merge_pass.cc:307]: I know, but this pass might mess up and influence the follow up passes.

File: paddle/fluid/framework/ir/multi_batch_merge_pass.cc (hunk 12 above)
Comment (panyx0718) [reply paddle/fluid/framework/ir/multi_batch_merge_pass.cc:314]: any doc about the assumptions of this pass?

File: python/paddle/fluid/tests/unittests/dist_mnist.py (hunk 30 above)
Comment (panyx0718) [reply python/paddle/fluid/tests/unittests/dist_mnist.py:98]: so this is no longer train reader?

File: python/paddle/fluid/tests/unittests/test_dist_mnist_batch_merge.py (hunk 31 above)
Comment (panyx0718) [reply python/paddle/fluid/tests/unittests/test_dist_mnist_batch_merge.py:63]: Why...does that mean a bug?

**Review by panyx0718 — APPROVED** [review]
Please add all necessary tests in follow ups...

**Review by qingqing01 — APPROVED** [review]
